### ISSUE TITLE
Extend composite compliant testing to backward formulas

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#69530]
## 🚀 Feature

"Composite compliant testing" checks that a composite operation works for Tensor-subclass-like Tensors (aka Tensor subclasses and Tensors without storage, like meta tensors). There are a class of backward formulas that are incompatible with Tensor subclasses due to the presence of in-place operations or .data_ptr() calls.

For example, consider the following:
```
torch.zeros(input.sizes(), grad.options()).diagonal().copy_(input)
```
If `input` is a Tensor subclass, but `grad` is a normal Tensor, then this operation will return us a regular Tensor as the gradient, not a Tensor subclass! In the past, both vmap and NestedTensor have faced this problem.

## Motivation

As Tensor subclasses become more common, we want more of PyTorch to work with them.

## Pitch

We should:
1. extend composite compliant testing to backward formulas
2. extend composite compliant testing to catch in-place operations that are incompatible with tensor subclasses.

## Alternatives

n/a
